File: Backend/catalog-service/src/middleware/tenant.js
// Middleware для извлечения tenant_id из URL и установки контекста

export const tenantMiddleware = (req, res, next) => {
  try {
    // Извлекаем tenant_id из различных источников
    let tenantId = null
    let subdomain = null

    // 1. Из заголовка X-Tenant-ID (приоритет)
    if (req.headers['x-tenant-id']) {
      tenantId = req.headers['x-tenant-id']
    }
    // 2. Из заголовка X-Tenant (legacy)
    else if (req.headers['x-tenant']) {
      tenantId = req.headers['x-tenant']
    }
    // 3. Из поддомена (например, shop1.marketplace.com)
    else if (req.headers.host) {
      subdomain = req.headers.host.split('.')[0]
      req.subdomain = subdomain
    }
    // 4. Из URL пути (например, /api/market/shop1/products)
    else if (req.path && req.path.includes('/market/')) {
      const pathParts = req.path.split('/')
      const marketIndex = pathParts.indexOf('market')
      if (marketIndex !== -1 && pathParts[marketIndex + 1]) {
        subdomain = pathParts[marketIndex + 1]
        req.subdomain = subdomain
      }
    }

    // Устанавливаем tenant_id или используем default для compatibility
    req.tenantId = tenantId || subdomain || 'default'
    req.tenant = { 
      id: req.tenantId,
      subdomain: subdomain || req.tenantId
    }

    // Добавляем флаг для проверки наличия tenant context
    req.hasTenantContext = !!tenantId || !!subdomain

    next()
  } catch (error) {
    console.error('Tenant middleware error:', error)
    res.status(500).json({ 
      error: 'Internal server error',
      message: 'Failed to process tenant context'
    })
  }
}

// Middleware для обязательной проверки tenant context
export const requireTenant = (req, res, next) => {
  if (!req.hasTenantContext && req.tenantId === 'default') {
    return res.status(400).json({
      error: 'Bad Request',
      message: 'Tenant context is required. Please provide X-Tenant-ID header or access via tenant subdomain.'
    })
  }
  next()
}

// Функция для получения tenant_id из БД по поддомену (для будущего использования с БД)
export async function getTenantBySubdomain(subdomain) {
  // TODO: Implement database query when PostgreSQL connection is added
  // For now, return mock data or subdomain as tenantId
  return {
    id: subdomain,
    subdomain: subdomain,
    name: `${subdomain} Marketplace`,
    status: 'active',
    tier: 'standard'
  }
}

// Middleware для загрузки полной информации о tenant
export const loadTenantInfo = () => async (req, res, next) => {
  try {
    if (req.subdomain && (!req.tenantId || req.tenantId === req.subdomain)) {
      const tenant = await getTenantBySubdomain(req.subdomain)
      
      if (!tenant) {
        return res.status(404).json({
          error: 'Not Found',
          message: `Tenant with subdomain '${req.subdomain}' not found or is not active`
        })
      }
      
      req.tenantId = tenant.id
      req.tenant = tenant
    }
    
    next()
  } catch (error) {
    console.error('Load tenant info error:', error)
    res.status(500).json({
      error: 'Internal server error',
      message: 'Failed to load tenant information'
    })
  }
}

File: Backend/auth-service/internal/middleware/tenant.go
package middleware

import (
	"context"
	"net/http"
	"strings"
)

// TenantContextKey is the key for tenant context
type TenantContextKey string

const (
	TenantIDKey     TenantContextKey = "tenantID"
	TenantKey       TenantContextKey = "tenant"
	SubdomainKey    TenantContextKey = "subdomain"
	HasTenantKey    TenantContextKey = "hasTenant"
)

// Tenant represents tenant information
type Tenant struct {
	ID        string `json:"id"`
	Subdomain string `json:"subdomain"`
	Name      string `json:"name"`
	Status    string `json:"status"`
	Tier      string `json:"tier"`
}

// TenantMiddleware extracts tenant ID from various sources
func TenantMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var tenantID string
		var subdomain string
		hasTenant := false

		// 1. Try X-Tenant-ID header (highest priority)
		if tid := r.Header.Get("X-Tenant-ID"); tid != "" {
			tenantID = tid
			hasTenant = true
		} else if tid := r.Header.Get("X-Tenant"); tid != "" {
			// 2. Try X-Tenant header (legacy)
			tenantID = tid
			hasTenant = true
		} else if host := r.Host; host != "" {
			// 3. Extract from subdomain
			parts := strings.Split(host, ".")
			if len(parts) > 0 {
				subdomain = parts[0]
				hasTenant = true
			}
		} else if strings.Contains(r.URL.Path, "/market/") {
			// 4. Extract from URL path (e.g., /market/shop1/products)
			pathParts := strings.Split(r.URL.Path, "/")
			for i, part := range pathParts {
				if part == "market" && i+1 < len(pathParts) {
					subdomain = pathParts[i+1]
					hasTenant = true
					break
				}
			}
		}

		// Default to "default" if no tenant context found
		if tenantID == "" && subdomain != "" {
			tenantID = subdomain
		} else if tenantID == "" {
			tenantID = "default"
		}

		// Create tenant object
		tenant := &Tenant{
			ID:        tenantID,
			Subdomain: subdomain,
		}

		// Add to context
		ctx := context.WithValue(r.Context(), TenantIDKey, tenantID)
		ctx = context.WithValue(ctx, TenantKey, tenant)
		ctx = context.WithValue(ctx, SubdomainKey, subdomain)
		ctx = context.WithValue(ctx, HasTenantKey, hasTenant)

		// Call next handler with updated context
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// RequireTenant middleware ensures tenant context exists
func RequireTenant(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		hasTenant := r.Context().Value(HasTenantKey)
		tenantID := r.Context().Value(TenantIDKey)

		if hasTenant == false || tenantID == "default" {
			http.Error(w, `{"error":"Bad Request","message":"Tenant context is required"}`, http.StatusBadRequest)
			return
		}

		next.ServeHTTP(w, r)
	})
}

// GetTenantID retrieves tenant ID from context
func GetTenantID(ctx context.Context) string {
	if tid, ok := ctx.Value(TenantIDKey).(string); ok {
		return tid
	}
	return "default"
}

// GetTenant retrieves tenant from context
func GetTenant(ctx context.Context) *Tenant {
	if tenant, ok := ctx.Value(TenantKey).(*Tenant); ok {
		return tenant
	}
	return &Tenant{ID: "default", Subdomain: "default"}
}

// GetSubdomain retrieves subdomain from context
func GetSubdomain(ctx context.Context) string {
	if subdomain, ok := ctx.Value(SubdomainKey).(string); ok {
		return subdomain
	}
	return ""
}

// HasTenantContext checks if tenant context exists
func HasTenantContext(ctx context.Context) bool {
	if has, ok := ctx.Value(HasTenantKey).(bool); ok {
		return has
	}
	return false
}
